proto/backend.proto

```diff
@@ -306,3 +306,160 @@ message StreamPacket {
+  rpc MutateAdmission(AdmissionRequest) returns (StorageResponse);
+
+  // Convert a resource to a new version
+  rpc ConvertObject(ConversionRequest) returns (StorageResponse);
+}
+
+message InstanceSettingsAdmissionRequest {
+  // NOTE: this may not include app or datasource instance settings depending on the request
+  PluginContext pluginContext = 1;
+
+  // The requested operation
+  StorageOperation operation = 2;
+
+  // Requested app instance state (not yet saved)
+  AppInstanceSettings appInstanceSettings = 3;
+
+  // Requested data source instance state (not yet saved)
+  DataSourceInstanceSettings dataSourceInstanceSettings = 4;
+}
+
+message InstanceSettingsResponse {
+  // Allowed indicates whether or not the admission request was permitted.
+  bool allowed = 1;
+
+  // Result contains extra details into why an admission request was denied.
+  // This field IS NOT consulted in any way if "Allowed" is "true".
+  // +optional
+  StatusResult result = 2;
+
+  // warnings is a list of warning messages to return to the requesting API client.
+  // Warning messages describe a problem the client making the API request should correct or be aware of.
+  // Limit warnings to 120 characters if possible.
+  // Warnings over 256 characters and large numbers of warnings may be truncated.
+  // +optional
+  repeated string warnings = 3;
+
+  // Valid app instance settings if they exist in the request
+  AppInstanceSettings appInstanceSettings = 4;
+
+  // Valid datasource instance settings if they exist in the request
+  DataSourceInstanceSettings dataSourceInstanceSettings = 5;
+}
+
+// AdmissionRequest contains information from a kubernetes Admission request and decoded object(s).
+// See: https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/admission/types.go#L41
+// And: https://github.com/grafana/grafana-app-sdk/blob/main/resource/admission.go#L14
+// NOTE: this does not include a plugin context
+message AdmissionRequest {
+  // NOTE: this may not include app or datasource instance settings depending on the request
+  PluginContext pluginContext = 1;
+
+  // The requested operation
+  StorageOperation operation = 2;
+  
+  // Object is the object in the request.  This includes the full metadata envelope.
+  // The Group+Version+Kind will be included in the payload
+  bytes object_bytes = 3;
+
+  // OldObject is the object as it currently exists in storage. This includes the full metadata envelope.
+  bytes old_object_bytes = 4;
+}
+
+// ConversionRequest supports converting an object from on version to another 
+message ConversionRequest {
+  // NOTE: this context exists because most of the middleware depends on it
+  // The format conversions most likely do not depend on instance settings passed with the context
+  PluginContext pluginContext = 1;
+
+  enum ObjectEnvelope {
+    // Kubernetes style resource
+    RESOURCE = 0;
+    // Full query request
+    QUERY_DATA_REQUEST = 1;
```

Comment:
When would a query request go through version conversion? Do aggregated API servers get admitting/validating/conversion calls?

Reply:
Queries are challenging!  The query needs to continue to take any query payload supported in its apiversion.  Unlike "normal" resources, queries are saved in objects that are owned by the apiserver defining the query behavior.  They are saved in dashboards, or alert definitions etc.

So we need *some* method to explicitly convert queries from one version to another